File: 1.1_is_unique/is_unique.php
<?php
/**
 * This is the solution for the Is unique problem, in PHP
 * 
 * @author Ruvart, <contacto@ruvart.com>
 */

declare(strict_types=1);

/**
 * If the string length is less than 2 return true.
 * 
 * Transform the string in to an array of chars, then create a second array with only the unique values of the array.
 * If the sizes of both arrays are equal, then return true, if not return false.
 * 
 * @param String $str
 * @return Bool
 */
function isUnique_1 ( String $str ): Bool {
    $str_length = strlen($str);

    if ( $str_length < 2 ) {
        return true;
    }

    $str_array        = str_split($str);
    $str_array_unique = array_unique($str_array);

    if ( $str_length != count($str_array_unique) ) {
        return false;
    }

    return true;
}


/**
 * If the string length is less than 2 return true.
 * 
 * Transform the string in to an array of chars, then for each char checks if it is set as key in the array of "chars found", if found then return false,
 * if not then add this char in to the array of "chars found" as a new key (associative arrayu)
 * 
 * @param String $str
 * @return Bool
 */
function isUnique_2 ( String $str ): Bool {
    $str_length  = strlen($str);
    $encontrados = [];

    if ( $str_length < 2 ) {
        return true;
    }

    $str_array = str_split($str);

    foreach ( $str_array as $char )  {
        if ( isset($encontrados[$char]) ) {
            return false;
        }

        $encontrados[$char] = true;
    }

    return true;
}


/**
 * If the string length is less than 2 return true.
 * 
 * Transform the string in to an array of chars, then search each char on an array of "chars found", if found then return false,
 * if not then add this char in to the array of "chars found" as a new value
 * 
 * @param String $str
 * @return Bool
 */
function isUnique_3 ( String $str ): Bool {
    $str_length  = strlen($str);
    $encontrados = [];

    if ( $str_length < 2 ) {
        return true;
    }

    $str_array = str_split($str);

    foreach ( $str_array as $char )  {
        if ( in_array($char,$encontrados) ) {
            return false;
        }

        $encontrados[] = $char;
    }

    return true;
}


/**
 * If the string length is less than 2 return true.
 * 
 * Transform the string in to an array of chars, then flip the arrar to use the values as keys so it will remain only 1 of each different char then compare the length of the result array with the string,
 * If the length is the same return true, if not return false
 * 
 * @param String $str
 * @return Bool
 */
function isUnique_4 ( String $str ): Bool {
    $str_length  = strlen($str);

    if ( $str_length < 2 ) {
        return true;
    }

    $str_array      = str_split($str);
    $str_array_flip = array_flip($str_array);

    if ( $str_length != count($str_array_flip) ) {
        return false;
    }

    return true;
}


/**
 * If the string length is less than 2 return true.
 * 
 * For each char in the string search for the same in the substribg from the position of this char to the end of the string
 * 
 * @param String $str
 * @return Bool
 */
function isUnique_5 ( String $str ): Bool {
    $str_length  = strlen($str);

    if ( $str_length < 2 ) {
        return true;
    }

    for ( $x = 1; $x < $str_length; $x++ ) {
        if ( strpos(substr($str,$x,$str_length-$x),$str[$x-1]) !== false ) {
            return false;
        }
    }

    return true;
}


/**
 * @param String $str
 * @return Bool
 */
function isUnique_6 ( String $str ): Bool {
    $str_length  = strlen($str);

    if ( $str_length < 2 ) {
        return true;
    }

    $str_array = str_split($str);
    sort($str_array);

    for ( $x = 1; $x < count($str_array); $x++ ) {
        if ( $str_array[$x-1] == $str_array[$x] ) {
            return false;
        }
    }

    return true;
}


$str_1 = "Neque porro quisquam est qui dolorem ipsum quia dolor sit amet, consectetur, adipisci velit";
$str_2 = "abcdefghijklmnopqrstuvwxyz ";


echo "isUnique_1 - (" . strlen($str_1) . ") \"" . $str_1 . "\":\n";
$ini    = microtime(true);
$result = isUnique_1($str_1);
$fin    = microtime(true);
$time   = round(($fin - $ini) * 1000);
if ( $result ) {
    echo "Son caracteres únicos.\t Tiempo (milisec): " . $time . "\n";
}
else {
    echo "No son caracteres únicos.\t Tiempo (milisec): " . $time . "\n";
}
echo "<br><br>\n\n";

echo "isUnique_1 - (" . strlen($str_2) . ") \"" . $str_2 . "\":\n";
$ini    = microtime(true);
$result = isUnique_1($str_2);
$fin    = microtime(true);
$time   = round(($fin - $ini) * 1000);
if ( $result ) {
    echo "Son caracteres únicos.\t Tiempo (milisec): " . $time . "\n";
}
else {
    echo "No son caracteres únicos.\t Tiempo (milisec): " . $time . "\n";
}
echo "<br><br>\n\n";


echo "isUnique_2 - (" . strlen($str_1) . ") \"" . $str_1 . "\":\n";
$ini    = microtime(true);
$result = isUnique_2($str_1);
$fin    = microtime(true);
$time   = round(($fin - $ini) * 1000);
if ( $result ) {
    echo "Son caracteres únicos.\t Tiempo (milisec): " . $time . "\n";
}
else {
    echo "No son caracteres únicos.\t Tiempo (milisec): " . $time . "\n";
}
echo "<br><br>\n\n";

echo "isUnique_2 - (" . strlen($str_2) . ") \"" . $str_2 . "\":\n";
$ini    = microtime(true);
$result = isUnique_2($str_2);
$fin    = microtime(true);
$time   = round(($fin - $ini) * 1000);
if ( $result ) {
    echo "Son caracteres únicos.\t Tiempo (milisec): " . $time . "\n";
}
else {
    echo "No son caracteres únicos.\t Tiempo (milisec): " . $time . "\n";
}
echo "<br><br>\n\n";


echo "isUnique_3 - (" . strlen($str_1) . ") \"" . $str_1 . "\":\n";
$ini    = microtime(true);
$result = isUnique_3($str_1);
$fin    = microtime(true);
$time   = round(($fin - $ini) * 1000);
if ( $result ) {
    echo "Son caracteres únicos.\t Tiempo (milisec): " . $time . "\n";
}
else {
    echo "No son caracteres únicos.\t Tiempo (milisec): " . $time . "\n";
}
echo "<br><br>\n\n";

echo "isUnique_3 - (" . strlen($str_2) . ") \"" . $str_2 . "\":\n";
$ini    = microtime(true);
$result = isUnique_3($str_2);
$fin    = microtime(true);
$time   = round(($fin - $ini) * 1000);
if ( $result ) {
    echo "Son caracteres únicos.\t Tiempo (milisec): " . $time . "\n";
}
else {
    echo "No son caracteres únicos.\t Tiempo (milisec): " . $time . "\n";
}
echo "<br><br>\n\n";


echo "isUnique_4 - (" . strlen($str_1) . ") \"" . $str_1 . "\":\n";
$ini    = microtime(true);
$result = isUnique_4($str_1);
$fin    = microtime(true);
$time   = round(($fin - $ini) * 1000);
if ( $result ) {
    echo "Son caracteres únicos.\t Tiempo (milisec): " . $time . "\n";
}
else {
    echo "No son caracteres únicos.\t Tiempo (milisec): " . $time . "\n";
}
echo "<br><br>\n\n";

echo "isUnique_4 - (" . strlen($str_2) . ") \"" . $str_2 . "\":\n";
$ini    = microtime(true);
$result = isUnique_4($str_2);
$fin    = microtime(true);
$time   = round(($fin - $ini) * 1000);
if ( $result ) {
    echo "Son caracteres únicos.\t Tiempo (milisec): " . $time . "\n";
}
else {
    echo "No son caracteres únicos.\t Tiempo (milisec): " . $time . "\n";
}
echo "<br><br>\n\n";


echo "isUnique_5 - (" . strlen($str_1) . ") \"" . $str_1 . "\":\n";
$ini    = microtime(true);
$result = isUnique_5($str_1);
$fin    = microtime(true);
$time   = round(($fin - $ini) * 1000);
if ( $result ) {
    echo "Son caracteres únicos.\t Tiempo (milisec): " . $time . "\n";
}
else {
    echo "No son caracteres únicos.\t Tiempo (milisec): " . $time . "\n";
}
echo "<br><br>\n\n";

echo "isUnique_5 - (" . strlen($str_2) . ") \"" . $str_2 . "\":\n";
$ini    = microtime(true);
$result = isUnique_5($str_2);
$fin    = microtime(true);
$time   = round(($fin - $ini) * 1000);
if ( $result ) {
    echo "Son caracteres únicos.\t Tiempo (milisec): " . $time . "\n";
}
else {
    echo "No son caracteres únicos.\t Tiempo (milisec): " . $time . "\n";
}
echo "<br><br>\n\n";


echo "isUnique_6 - (" . strlen($str_1) . ") \"" . $str_1 . "\":\n";
$ini    = microtime(true);
$result = isUnique_6($str_1);
$fin    = microtime(true);
$time   = round(($fin - $ini) * 1000);
if ( $result ) {
    echo "Son caracteres únicos.\t Tiempo (milisec): " . $time . "\n";
}
else {
    echo "No son caracteres únicos.\t Tiempo (milisec): " . $time . "\n";
}
echo "<br><br>\n\n";

echo "isUnique_6 - (" . strlen($str_2) . ") \"" . $str_2 . "\":\n";
$ini    = microtime(true);
$result = isUnique_6($str_2);
$fin    = microtime(true);
$time   = round(($fin - $ini) * 1000);
if ( $result ) {
    echo "Son caracteres únicos.\t Tiempo (milisec): " . $time . "\n";
}
else {
    echo "No son caracteres únicos.\t Tiempo (milisec): " . $time . "\n";
}
echo "<br><br>\n\n";

File: 1.1_is_unique/is_unique.py
"""This is the solution for the Is unique problem, in Python
"""

__author__ = "Ruvart"
__email__ = "contacto@ruvart.com"


from collections import Counter
import time


def isUnique_1(text: str) -> bool:
    """If the length of the string is < 2 return True.
    Uses Counter to determinehow many times each character appear in the string, then filter theh result to get only the ones that repeat once.
    If the length of the result is differente than the length of the string, then return false, else true"""

    str_len = len(text)

    if str_len < 2:
        return True

    c_str = Counter(text)

    unique_str = [key for key in c_str.keys() if c_str[key] == 1]

    if len(unique_str) != str_len:
        return False

    return True


def isUnique_2(text: str) -> bool:
    """If the length of the string is < 2 return True.
    Sorts the string, then for each character checks if it is the same as the last one, if it is equal return false, else true"""

    str_len = len(text)

    if str_len < 2:
        return True

    sorted_str = sorted(text)
    for x in range(1,str_len):
        if sorted_str[x-1] == sorted_str[x]:
            return False
    
    return True


def isUnique_3(text: str) -> bool:
    """If the length of the string is < 2 return True.
    For each char in the string search for the same in the substring from the position of this char to the end of the string"""
    
    str_len = len(text)

    if str_len < 2:
        return True

    for x in range(str_len):
        if text.find(text[x],x+1) != -1:
            return False
    
    return True




str_1 = "Neque porro quisquam est qui dolorem ipsum quia dolor sit amet, consectetur, adipisci velit"
str_2 = "abcdefghijklmnopqrstuvwxyz "

print(f"isUnique_1 - ({len(str_1)}) \"{str_1}\":")
t_ini = time.process_time_ns()/1000000
result = isUnique_1(str_1)
t_fin = time.process_time_ns()/1000000
if result:
    print(f"Son caracteres únicos.\t Tiempo (milisec): {t_fin - t_ini}\n")
else:
    print(f"No son caracteres únicos.\t Tiempo (milisec): {t_fin - t_ini}\n")

print(f"isUnique_1 - ({len(str_2)}) \"{str_2}\":")
t_ini = time.process_time_ns()/1000000
result = isUnique_1(str_2)
t_fin = time.process_time_ns()/1000000
if result:
    print(f"Son caracteres únicos.\t Tiempo (milisec): {t_fin - t_ini}\n")
else:
    print(f"No son caracteres únicos.\t Tiempo (milisec): {t_fin - t_ini}\n")



print(f"isUnique_2 - ({len(str_1)}) \"{str_1}\":")
t_ini = time.process_time_ns()/1000000
result = isUnique_2(str_1)
t_fin = time.process_time_ns()/1000000
if result:
    print(f"Son caracteres únicos.\t Tiempo (milisec): {t_fin - t_ini}\n")
else:
    print(f"No son caracteres únicos.\t Tiempo (milisec): {t_fin - t_ini}\n")

print(f"isUnique_2 - ({len(str_2)}) \"{str_2}\":")
t_ini = time.process_time_ns()/1000000
result = isUnique_2(str_2)
t_fin = time.process_time_ns()/1000000
if result:
    print(f"Son caracteres únicos.\t Tiempo (milisec): {t_fin - t_ini}\n")
else:
    print(f"No son caracteres únicos.\t Tiempo (milisec): {t_fin - t_ini}\n")



print(f"isUnique_3 - ({len(str_1)}) \"{str_1}\":")
t_ini = time.process_time_ns()/1000000
result = isUnique_3(str_1)
t_fin = time.process_time_ns()/1000000
if result:
    print(f"Son caracteres únicos.\t Tiempo (milisec): {t_fin - t_ini}\n")
else:
    print(f"No son caracteres únicos.\t Tiempo (milisec): {t_fin - t_ini}\n")

print(f"isUnique_3 - ({len(str_2)}) \"{str_2}\":")
t_ini = time.process_time_ns()/1000000
result = isUnique_3(str_2)
t_fin = time.process_time_ns()/1000000
if result:
    print(f"Son caracteres únicos.\t Tiempo (milisec): {t_fin - t_ini}\n")
else:
    print(f"No son caracteres únicos.\t Tiempo (milisec): {t_fin - t_ini}\n")